src/rust/rust-proto/src/graplinc/grapl/api/client.rs

```diff
@@ -0,0 +1,381 @@
+        }
+
+        async fn connect_with_healthcheck(
+            configuration: ClientConfiguration,
+            healthcheck_timeout: Duration,
+            healthcheck_polling_interval: Duration,
+        ) -> Result<Self, ClientError> {
+            let client = Client::connect_with_healthcheck(
+                configuration,
+                Self::SERVICE_NAME,
+                healthcheck_timeout,
+                healthcheck_polling_interval,
+            )
+            .await?;
+
+            Ok(Self::with_client(client))
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Client<C>
+where
+    C: Connectable + Clone + Debug,
+{
+    configuration: ClientConfiguration,
+    proto_client: C,
+    client_executor: Executor,
+}
+
+impl<C> Client<C>
+where
+    C: Connectable + Clone + Debug,
+{
+    pub(crate) async fn connect(configuration: ClientConfiguration) -> Result<Self, ClientError> {
+        let connect_executor = Executor::new(ExecutorConfig::new(configuration.connect_timeout));
+        let endpoint = configuration.endpoint()?;
+        let proto_client = connect_executor
+            .spawn(
+                configuration
+                    .connect_backoff_strategy()
+                    .take(configuration.connect_retries),
+                || {
+                    let endpoint = endpoint.clone();
+                    async move { C::connect(endpoint).await.map_err(ClientError::from) }
+                },
+            )
+            .await?;
+
+        let client_executor = Executor::new(ExecutorConfig::new(configuration.executor_timeout));
+
+        Ok(Self {
+            configuration,
+            proto_client,
+            client_executor,
+        })
+    }
+
+    pub(crate) async fn connect_with_healthcheck(
+        configuration: ClientConfiguration,
+        service_name: &'static str,
+        healthcheck_timeout: Duration,
+        healthcheck_polling_interval: Duration,
+    ) -> Result<Self, ClientError> {
+        let endpoint = configuration.endpoint()?;
+        HealthcheckClient::wait_until_healthy(
+            endpoint,
+            service_name,
+            healthcheck_timeout,
+            healthcheck_polling_interval,
+        )
+        .await?;
+
+        Self::connect(configuration).await
+    }
+
+    pub(crate) async fn execute<PT, NT, PU, NU, P, F, R>(
+        &self,
+        request: NT,
+        retry_predicate: P,
+        max_retries: usize,
+        grpc_call: F,
+    ) -> Result<NU, ClientError>
+    where
+        PT: prost::Message + From<NT> + Clone,
+        NT: ProtobufSerializable<ProtobufMessage = PT> + TryFrom<PT>,
+        PU: prost::Message,
+        NU: ProtobufSerializable<ProtobufMessage = PU> + TryFrom<PU>,
+        P: Fn(&tonic::Status) -> bool,
+        F: FnMut(C, tonic::Request<PT>) -> R + Clone,
+        R: Future<Output = Result<tonic::Response<PU>, tonic::Status>>,
+        ClientError: From<<NU as TryFrom<PU>>::Error>,
```

Comment:
It's a little hard for me to reason about the names of these type vars; would you be amenable to e.g.

ProtoRequest, NativeRequest, ProtoResponse, NativeResponse 
ProtoReq, NativeResp
or even PReq, NResp? 

If you'd like to keep them short, that's fine, but could you leave a comment about what each of the [P,N][T,U] represents? 

i'm assuming PT/PU is just from T, U in the alphabet?

Reply:
yeah the idea here is we have a `P` for "proto" and `N` for "native" and the normal notation for a function of type `F` is `F<T, U>` which takes domain `T` to range `U`. I can make this more verbose.